klippy/extras/display/menu.py

```diff
@@ -0,0 +1,1137 @@
+
+    def is_accepted(self, item):
+        return super(MenuList, self).is_accepted(item) and type(item) is not MenuCard
+
+    def _names_aslist(self):
+        return self._aslist_splitlines(self.items)
+
+    def _lookup_item(self, item):
+        if isinstance(item, str) and ',' in item:
+            item = MenuGroup(self._menu, {'name':'Group', 'items': item}, ',')
+        return super(MenuList, self)._lookup_item(item)
+
+    def update_items(self):
+        super(MenuList, self).update_items()
+        for item in self._items:
+            if isinstance(item, MenuGroup) and not item.is_editing():
+                item.update_items()
+
+    def get_enter_gcode(self):
+        return self._enter_gcode
+
+    def get_leave_gcode(self):
+        return self._leave_gcode
+
+class MenuVSDCard(MenuList):
+    def __init__(self, menu, config):
+        super(MenuVSDCard, self).__init__(menu, config)
+
+    def _populate_files(self):
+        sdcard = self._menu.objs['virtual_sdcard']
+        if sdcard is not None:
+            files = sdcard.get_file_list()
+            for fname, fsize in files:
+                gcode = [
+                    'M23 /%s' % str(fname)
+                ]
+                self.append_item(MenuCommand(self._menu, {'name': '%s' % str(fname), 'cursor':'+', 'gcode': "\n".join(gcode)}))
+
+    def populate_items(self):
+        super(MenuVSDCard, self).populate_items()
+        self._populate_files()
+
+class MenuCard(MenuGroup):
+    def __init__(self, menu, config):
+        super(MenuCard, self).__init__(menu, config)
+        self.content = config.get('content')
+
+    def _names_aslist(self):
+        return self._aslist_splitlines(self.items)
+
+    def _content_aslist(self):
+        return filter(None, [self._strip_quotes(s) for s in self._aslist_splitlines(self.content)])
+
+    def update_items(self):
+        self._items = self._allitems[:]
+        for item in self._items:
+            if isinstance(item, MenuGroup) and not item.is_editing():
+                item.update_items()
+
+    def _lookup_item(self, item):
+        if isinstance(item, str) and ',' in item:
+            item = MenuCycler(self._menu, {'name':'Cycler', 'items': item}, ',')
+        return super(MenuCard, self)._lookup_item(item)
+
+    def render_content(self, eventtime):
+        if self.selected is not None:
+            self.selected = (self.selected % len(self)) if len(self) > 0 else None
+
+        items = []
+        for i, item in enumerate(self):
+            name = ''
+            if item.is_enabled():
+                item.heartbeat(eventtime)
+                name = self._render_item(item, (i == self.selected), True)
+            items.append(name)
+        lines = []
+        for line in self._content_aslist():
+            try:
+                lines.append(str(line).format(*items))
+            except Exception as e:
+                logging.error('Card rendering exception: '+ str(e))
```

Comment:
It would be preferable to use logging.exception("Parameter update error:") instead of manually calling logging.error().